File: Deque/deque.py
class Deque:
    def __init__(self):
        self.cap=4
        self.size=0
        self.front=0
        self.data=[None]*self.cap
        
    def resize(self):
        self.cap*=2
        newData= [None]*self.cap
        
        for i in range(self.size):
            newData[i]= self.data[(i+self.cap)%self.cap]
            
        self.front=0
        
    def push_back(self, val):
        if self.size==self.cap:
            self.resize()
        rear= (self.front+self.size)%self.cap
        self.data[rear]= val
        
        self.size+=1
        
    def push_front(self, val):
        if self.size==self.cap:
            self.resize()
        self.front= (self.front-1+self.cap)%self.cap
        self.data[self.front]= val
        
        self.size+=1
        
    def pop_back(self):
        if self.empty():
            raise IndexError("Empty deque")
        rear= (self.front+self.size-1)%self.cap
        val= self.data[rear]
        self.data[rear]= None
        self.size-=1
        
        return val
    
    def pop_front(self):
        if self.empty():
            raise IndexError("Empty deque")
        self.front= (self.front-1)%self.cap
        val = self.data[self.front]
        self.data[self.front]= None
        self.size-=1
        
        return val
    
    def empty(self):
        return self.size==0
    
    
        
            

File: Deque/deque.cpp
template <typename T>
class Deque
{
    private:
        T* data;
        int cap;
        int size;
        int front;

        void resize()
        {
            cap*=2;
            T* newData= new T[cap];

            for (int i=0; i<size; i++)
            {
                newData[i]= data[(front+i)%cap];
            }

            delete[] data;
            data= newData;

            front= 0;
        }

    public:
        Deque()
        {
            cap= 4;
            size=0;
            data= new T[cap];
            frot=0;
        }

        ~Deque()
        {
            delete[] data;
        }

        void push_back(const T& val)
        {
            if (size==cap)
            {
                resize();
            }
            int rear= (front+size)%cap;
            data[rear]= val;
            size++;
        }

        void push_front(const T& val)
        {
            if (size==cap)
            {
                resize();
            }
            front= (front-1+cap)%cap;
            data[front]= val;
            size++;
        }

        T pop_back()
        {
            if (empty())
            {
                throw std::index_error("Empty queue")
            }
            int rear= (front+size-1)%cap;
            T val= data[rear];
            size--;

            return val;
        }

        T pop_front()
        {
            if (empty())
            {
                throw std::index_error("Empty queue")
            }
            front= (front+1)%cap;
            T val= data[front];
            size--;

            return val;
        }

        bool empty()
        {
            return (size==0);
        }

};